src/trace-plugin-loader.ts

```diff
@@ -16,278 +16,485 @@
+ * function, as well as the version range of the module for which that file
+ * should be patched. (See ./plugin-types for the exact interface.)
  */
-export interface PluginLoaderConfig extends TraceAgentConfig {
-  plugins: {[pluginName: string]: string;};
-}
 
-interface InternalPatch<T> extends Patch<T> {
-  file: string;
-  module?: T;
+export interface PluginLoaderConfig extends TraceAgentConfig {
+  // An object which contains paths to files that should be loaded as plugins
+  // upon loading a module with a given name.
+  plugins: {[pluginName: string]: string};
 }
 
-interface InternalIntercept<T> extends Intercept<T> {
-  file: string;
-  module?: T;
+/**
+ * An interface representing configuration passed to the plugin loader, which
+ * includes TraceAgent configuration as well.
+ */
+export interface PluginLoaderSingletonConfig extends PluginLoaderConfig {
+  forceNewAgent_: boolean;
 }
 
-type InternalInstrumentation<T> = InternalPatch<T>|InternalIntercept<T>;
-
-// tslint:disable:no-any
-interface InternalPlugin {
-  file: string;
-  patches:
-      {[patchName: string]: {[file: string]: InternalInstrumentation<any>;}};
-  agent: TraceAgent;
+export interface ModulePluginWrapperOptions {
+  name: string;
+  path: string;
 }
-// tslint:enable:no-any
 
-interface PluginStore {
-  [pluginName: string]: InternalPlugin;
+export interface CorePluginWrapperOptions {
+  children: ModulePluginWrapperOptions[];
 }
 
-// type guards
+/**
+ * An interface that represents a wrapper around user-provided plugin objects.
+ */
+export interface PluginWrapper {
+  /**
+   * Returns whether the given version of the module is supported by this
+   * plugin. This may load the plugin into memory.
+   * @param version A semver version string.
+   */
+  isSupported(version: string): boolean;
 
-function isPatch<T>(obj: Instrumentation<T>): obj is Patch<T> {
-  return !!(obj as Patch<T>).patch;
-}
+  /**
+   * Call unpatch methods when they were provided.
+   */
+  unapplyAll(): void;
 
-function isIntercept<T>(obj: Instrumentation<T>): obj is Intercept<T> {
-  return !!(obj as Intercept<T>).intercept;
+  /**
+   * Applies this object's underlying plugin patches to a file, returning the
+   * patched or intercepted value.
+   * @param moduleExports The module exports of the file.
+   * @param file The file path, relative to the module root.
+   * @param version The module version.
+   */
+  applyPlugin<T>(moduleExports: T, file: string, version: string): T;
 }
 
-function isInternalPatch<T>(obj: InternalInstrumentation<T>):
-    obj is InternalPatch<T> {
-  return !!(obj as InternalPatch<T>).patch;
-}
+/**
+ * A class that represents wrapper logic around a user-provided plugin object
+ * to be applied to a single module.
+ */
+export class ModulePluginWrapper implements PluginWrapper {
+  // Sentinel value to indicate that a plugin has not been loaded into memory
+  // yet.
+  private static readonly NOT_LOADED: Plugin = [];
```

Comment:
Would it make sense to export this from plugin-types.ts? Probably after naming it `NULL_PLUGIN` or something.

Reply:
I'm not sure how it might be used by end users. Its value is only used internally by a class that is not exposed to users.

Reply:
My comment was less about usages by end users but more about keeping the abstraction barrier. Here it is assumed that `Plugin` is actually an array. If plugin-types.ts exported `NULL_PLUGIN`, we wouldn't have to know if it's an array or an object or whatever.

Having said that, since other parts of code in this file assume that a plugin is an array, I don't think it buys much.

Reply:
`NOT_LOADED` is only used privately in `ModulePluginWrapper`; even the `Plugin` field that is initially assigned to it is inaccessible. (IOW even within this file, other classes don't know about it.) It might as well be a symbol -- the only reason it's not is to make type checking simple.